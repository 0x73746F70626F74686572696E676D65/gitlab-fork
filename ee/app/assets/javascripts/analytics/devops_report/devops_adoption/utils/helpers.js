/**
 * A helper function which accepts the enabledNamespaces,
 *
 * @param {Object} params the enabledNamespaces data, and the check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ enabledNamespaces, openModal }) => {
  if (openModal) {
    return false;
  } else if (!enabledNamespaces.length) {
    return true;
  }

  const anyPendingEnabledNamespaces = enabledNamespaces.some(
    (node) => node.latestSnapshot === null,
  );

  return anyPendingEnabledNamespaces;
};
