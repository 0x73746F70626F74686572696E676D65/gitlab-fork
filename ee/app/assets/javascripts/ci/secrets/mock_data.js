const mockSecretsData = [
  {
    name: 'runner-token-windows',
    labels: [],
    lastAccessed: '2024-03-19T20:55:08.551Z',
    createdAt: '2024-03-11T01:33:06.258Z',
  },
  {
    name: 'product/client-tokens',
    labels: ['env::staging', 'env::production'],
    lastAccessed: '2024-01-23T07:12:44.833Z',
    createdAt: '2024-01-22T08:04:26.024Z',
  },
  {
    name: 'product/server-tokens',
    labels: ['env::staging', 'env::production'],
    lastAccessed: '2024-01-22T07:12:44.833Z',
    createdAt: '2023-01-24T08:04:26.024Z',
  },
  {
    name: 'security-token-project2',
    labels: ['env::production', 'security'],
    lastAccessed: '2024-01-21T23:13:37.227Z',
    createdAt: '2024-01-19T11:59:43.516Z',
  },
  {
    name: 'personal-tokens',
    labels: [],
    lastAccessed: '2024-01-19T21:01:38.925Z',
    createdAt: '2024-01-17T05:51:54.602Z',
  },
  {
    name: 'token-deployment',
    labels: [],
    lastAccessed: '2024-01-16T23:13:31.548Z',
    createdAt: '2024-01-15T17:12:33.882Z',
  },
  {
    name: 'runner-token-mac',
    labels: [],
    lastAccessed: '2024-01-06T07:50:58.308Z',
    createdAt: '2023-01-15T16:35:38.326Z',
  },
  {
    name: 'aws-db-prod-credentials',
    labels: ['AWS'],
    lastAccessed: '2024-01-05T11:47:59.454Z',
    createdAt: '2024-01-04T10:37:41.135Z',
  },
  {
    name: 'runner-token-linux',
    labels: [],
    lastAccessed: '2023-12-30T16:41:08.551Z',
    createdAt: '2023-01-14T18:52:56.258Z',
  },
  {
    name: 'SSH-KEY-1',
    labels: [],
    lastAccessed: '2023-11-29T22:18:31.562Z',
    createdAt: '2023-11-30T01:29:14.394Z',
  },
  {
    name: 'group-ID',
    labels: ['shell'],
    lastAccessed: '2023-11-29T04:30:10.220Z',
    createdAt: '2023-01-21T17:34:33.544Z',
  },
  {
    name: 'security-token-project1',
    labels: ['security'],
    lastAccessed: '2023-11-28T04:30:10.220Z',
    createdAt: '2023-01-21T17:34:33.544Z',
  },
];

export const mockGroupSecretsData = mockSecretsData.map((secret, index) => ({
  key: `group_secret_${index}`,
  ...secret,
}));

export const mockProjectSecretsData = mockSecretsData.map((secret, index) => ({
  key: `project_secret_${index}`,
  ...secret,
}));
