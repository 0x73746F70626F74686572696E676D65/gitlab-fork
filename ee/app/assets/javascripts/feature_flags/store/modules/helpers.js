// eslint-disable-next-line import/prefer-default-export
export const parseFeatureFlagsParams = params => ({
  operations_feature_flags: {
    name: params.name,
    description: params.description,
    active: true,
    scopes_attributes: params.scopes.map(scope => ({
      environment_scope: scope.name,
      active: scope.active,
    })),
  },
});
