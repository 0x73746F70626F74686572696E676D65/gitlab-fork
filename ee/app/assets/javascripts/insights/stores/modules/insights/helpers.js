import { CHART_TYPES } from 'ee/insights/constants';
import { __ } from '~/locale';

const getGroupByValue = (queryParams) => {
  return queryParams.group_by || queryParams.params.group_by;
};

const getTypeValue = (queryParams) => {
  return queryParams.issuable_type || queryParams.params.issuable_type;
};

const getAxisTitle = (label) => {
  switch (label) {
    case 'day':
      return __('Days');
    case 'week':
      return __('Weeks');
    case 'month':
      return __('Months');
    case 'issue':
      return __('Issues');
    case 'merge_request':
      return __('Merge requests');
    default:
      return '';
  }
};

export const transformChartDataForGlCharts = (
  { type, query: queryParams },
  { labels, datasets },
) => {
  const formattedData = {
    xAxisTitle: getAxisTitle(getGroupByValue(queryParams)),
    yAxisTitle: getAxisTitle(getTypeValue(queryParams)),
    labels,
    datasets: [],
    seriesNames: [],
  };

  switch (type) {
    case CHART_TYPES.BAR:
      formattedData.datasets = [
        {
          name: 'all',
          data: labels.map((label, i) => [label, datasets[0].data[i]]),
        },
      ];
      break;
    case CHART_TYPES.STACKED_BAR:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: dataset.data,
        })),
      );
      break;
    case CHART_TYPES.LINE:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: labels.map((label, i) => [label, dataset.data[i]]),
        })),
      );

      break;
    default:
      formattedData.datasets = { all: labels.map((label, i) => [label, datasets[0].data[i]]) };
  }
  return formattedData;
};

export default {
  transformChartDataForGlCharts,
};
