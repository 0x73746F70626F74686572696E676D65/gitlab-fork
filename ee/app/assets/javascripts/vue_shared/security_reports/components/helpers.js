import { EMPTY_BODY_MESSAGE } from './constants';

/**
 * A helper function which validates the passed
 * in body string.
 *
 * It returns an empty string if the body has explicitly
 * been passed in as an empty string, a fallback
 * message if the body is null / undefined, else
 * it will return the original body string.
 *
 * @param {String} body the body message
 *
 * @return {String} the validated body message
 */
export const bodyWithFallBack = (body) => (body === '' ? '' : body || EMPTY_BODY_MESSAGE);

export const getCreatedIssueForVulnerability = (vulnerability) =>
  vulnerability.issue_links?.find((link) => link.link_type === 'created');

export const getDismissalTransitionForVulnerability = (vulnerability) => {
  const latestTransition = vulnerability.state_transitions?.at(-1);
  return latestTransition?.to_state.toLowerCase() === 'dismissed' ? latestTransition : null;
};
