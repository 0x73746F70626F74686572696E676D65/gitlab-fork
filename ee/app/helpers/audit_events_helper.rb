# frozen_string_literal: true

module AuditEventsHelper
  def human_text(details)
    return custom_message_for(details) if details[:custom_message]

    details.map { |key, value| select_keys(key, value) }.join(" ").humanize
  end

  def select_keys(key, value)
    if key =~ /^(author|target)_.*/
      ""
    else
      "#{key} <strong>#{value}</strong>"
    end
  end

  def custom_message_for(details)
    target_type = details[:target_type]
    val = details[:custom_message]
    target_type == 'Operations::FeatureFlag' ? val : val.tr('_', ' ')
  end
end
