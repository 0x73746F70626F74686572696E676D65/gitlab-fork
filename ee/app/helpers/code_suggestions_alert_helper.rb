# frozen_string_literal: true

module CodeSuggestionsAlertHelper
  def show_code_suggestions_alert?
    return false unless ::Gitlab::CurrentSettings.should_check_namespace_plan?
    return false if cookies[:code_suggestions_alert_dismissed] == 'true'
    return false if Feature.disabled?(:code_suggestions_alert, current_user)
    return true if current_user.blank?

    !current_user.code_suggestions_enabled?
  end
end
