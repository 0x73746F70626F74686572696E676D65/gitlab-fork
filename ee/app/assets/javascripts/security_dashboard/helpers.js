import isPlainObject from 'lodash/isPlainObject';
import { s__ } from '~/locale';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { ALL, BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';

const parseOptions = obj =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = projects =>
  projects.map(p => ({ id: p.id.split('/').pop(), name: p.name }));

export const initFirstClassVulnerabilityFilters = projects => {
  const filters = [
    {
      name: s__('SecurityReports|Status'),
      id: 'state',
      options: [
        { id: ALL, name: s__('VulnerabilityStatusTypes|All') },
        ...parseOptions(VULNERABILITY_STATES),
      ],
      selection: new Set([ALL]),
    },
    {
      name: s__('SecurityReports|Severity'),
      id: 'severity',
      options: [BASE_FILTERS.severity, ...parseOptions(SEVERITY_LEVELS)],
      selection: new Set([ALL]),
    },
    {
      name: s__('SecurityReports|Report type'),
      id: 'reportType',
      options: [BASE_FILTERS.report_type, ...parseOptions(REPORT_TYPES)],
      selection: new Set([ALL]),
    },
  ];

  if (Array.isArray(projects)) {
    filters.push({
      name: s__('SecurityReports|Project'),
      id: 'projectId',
      options: [BASE_FILTERS.project_id, ...mapProjects(projects)],
      selection: new Set([ALL]),
    });
  }

  return filters;
};

export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter(entry => entry !== null);
};

export default () => ({});
