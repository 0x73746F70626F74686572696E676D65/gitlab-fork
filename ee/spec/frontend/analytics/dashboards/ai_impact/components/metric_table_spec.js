import Vue from 'vue';
import VueApollo from 'vue-apollo';
import { GlTooltip } from '@gitlab/ui';
import {
  FLOW_METRICS,
  DORA_METRICS,
  VULNERABILITY_METRICS,
  AI_METRICS,
} from '~/analytics/shared/constants';
import { mountExtended } from 'helpers/vue_test_utils_helper';
import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';
import FlowMetricsQuery from 'ee/analytics/dashboards/ai_impact/graphql/flow_metrics.query.graphql';
import DoraMetricsQuery from 'ee/analytics/dashboards/ai_impact/graphql/dora_metrics.query.graphql';
import VulnerabilitiesQuery from 'ee/analytics/dashboards/ai_impact/graphql/vulnerabilities.query.graphql';
import AiMetricsQuery from 'ee/analytics/dashboards/ai_impact/graphql/ai_metrics.query.graphql';
import MetricTable from 'ee/analytics/dashboards/ai_impact/components/metric_table.vue';
import MetricTableCell from 'ee/analytics/dashboards/components/metric_table_cell.vue';
import TrendIndicator from 'ee/analytics/dashboards/components/trend_indicator.vue';
import { setLanguage } from 'jest/__helpers__/locale_helper';
import {
  mockDoraMetricsResponse,
  mockFlowMetricsResponse,
  mockVulnerabilityMetricsResponse,
  mockAiMetricsResponse,
} from '../helpers';
import { mockTableValues, mockTableLargeValues, mockAiMetricsValues } from '../mock_data';

const mockTypePolicy = {
  Query: { fields: { project: { merge: false }, group: { merge: false } } },
};

Vue.use(VueApollo);

describe('Metric table', () => {
  let wrapper;

  const namespace = 'test-namespace';
  const isProject = false;

  const createMockApolloProvider = ({
    flowMetricsRequest = mockFlowMetricsResponse(mockTableValues),
    doraMetricsRequest = mockDoraMetricsResponse(mockTableValues),
    vulnerabilityMetricsRequest = mockVulnerabilityMetricsResponse(mockTableValues),
    aiMetricsRequest = mockAiMetricsResponse(mockAiMetricsValues),
  } = {}) => {
    return createMockApollo(
      [
        [FlowMetricsQuery, flowMetricsRequest],
        [DoraMetricsQuery, doraMetricsRequest],
        [VulnerabilitiesQuery, vulnerabilityMetricsRequest],
        [AiMetricsQuery, aiMetricsRequest],
      ],
      {},
      {
        typePolicies: mockTypePolicy,
      },
    );
  };

  const createMockApolloProviderLargeValues = ({
    flowMetricsRequest = mockFlowMetricsResponse(mockTableLargeValues),
    doraMetricsRequest = mockDoraMetricsResponse(mockTableLargeValues),
    vulnerabilityMetricsRequest = mockVulnerabilityMetricsResponse(mockTableLargeValues),
    aiMetricsRequest = mockAiMetricsResponse(mockTableLargeValues),
  } = {}) => {
    return createMockApollo(
      [
        [FlowMetricsQuery, flowMetricsRequest],
        [DoraMetricsQuery, doraMetricsRequest],
        [VulnerabilitiesQuery, vulnerabilityMetricsRequest],
        [AiMetricsQuery, aiMetricsRequest],
      ],
      {},
      {
        typePolicies: mockTypePolicy,
      },
    );
  };

  const createWrapper = ({ props = {}, apolloProvider = createMockApolloProvider() } = {}) => {
    wrapper = mountExtended(MetricTable, {
      apolloProvider,
      propsData: {
        namespace,
        isProject,
        ...props,
      },
    });

    return waitForPromises();
  };

  const createLargeValuesWrapper = ({
    props = {},
    apolloProvider = createMockApolloProviderLargeValues(),
  } = {}) => {
    wrapper = mountExtended(MetricTable, {
      apolloProvider,
      propsData: {
        namespace,
        isProject,
        ...props,
      },
    });

    return waitForPromises();
  };

  const findTableRow = (rowTestId) => wrapper.findByTestId(rowTestId);
  const findMetricTableCell = (rowTestId) => findTableRow(rowTestId).findComponent(MetricTableCell);
  const findValueTableCells = (rowTestId) =>
    findTableRow(rowTestId).findAll(`[data-testid="ai-impact-table-value-cell"]`);
  const findTrendIndicator = (rowTestId) => findTableRow(rowTestId).findComponent(TrendIndicator);
  const findSkeletonLoaders = (rowTestId) =>
    wrapper.findAll(`[data-testid="${rowTestId}"] [data-testid="metric-skeleton-loader"]`);

  describe.each`
    identifier                                | name                                    | testId                                            | change  | hasValueTooltips
    ${DORA_METRICS.DEPLOYMENT_FREQUENCY}      | ${'Deployment frequency'}               | ${'ai-impact-metric-deployment-frequency'}        | ${1}    | ${false}
    ${DORA_METRICS.CHANGE_FAILURE_RATE}       | ${'Change failure rate'}                | ${'ai-impact-metric-change-failure-rate'}         | ${1}    | ${false}
    ${FLOW_METRICS.CYCLE_TIME}                | ${'Cycle time'}                         | ${'ai-impact-metric-cycle-time'}                  | ${-0.5} | ${false}
    ${FLOW_METRICS.LEAD_TIME}                 | ${'Lead time'}                          | ${'ai-impact-metric-lead-time'}                   | ${0}    | ${false}
    ${VULNERABILITY_METRICS.CRITICAL}         | ${'Critical vulnerabilities over time'} | ${'ai-impact-metric-vulnerability-critical'}      | ${-0.5} | ${false}
    ${AI_METRICS.CODE_SUGGESTIONS_USAGE_RATE} | ${'Code Suggestions usage'}             | ${'ai-impact-metric-code-suggestions-usage-rate'} | ${1}    | ${true}
  `('for the $identifier table row', ({ identifier, name, testId, change, hasValueTooltips }) => {
    describe('when loading data', () => {
      beforeEach(() => {
        createWrapper();
      });

      it('renders the metric name', () => {
        expect(findMetricTableCell(testId).props()).toEqual(
          expect.objectContaining({
            identifier,
            requestPath: namespace,
            isProject,
          }),
        );
      });

      it('renders a skeleton loader in each cell', () => {
        // Metric count + 1 for the trend indicator
        const loadingCellCount = Object.keys(mockTableValues[0]).length + 1;
        expect(findSkeletonLoaders(testId).length).toBe(loadingCellCount);
      });
    });

    describe('when the data fails to load', () => {
      beforeEach(() => {
        return createWrapper({
          apolloProvider: createMockApolloProvider({
            flowMetricsRequest: jest.fn().mockRejectedValue({}),
            doraMetricsRequest: jest.fn().mockRejectedValue({}),
            vulnerabilityMetricsRequest: jest.fn().mockRejectedValue({}),
            aiMetricsRequest: jest.fn().mockRejectedValue({}),
          }),
        });
      });

      it('emits `set-alerts` with the name of the failed metric', () => {
        expect(wrapper.emitted('set-alerts')).toHaveLength(1);
        expect(wrapper.emitted('set-alerts')[0][0].errors).toHaveLength(1);
        expect(wrapper.emitted('set-alerts')[0][0].errors[0]).toContain(name);
      });
    });

    describe('when the data is loaded', () => {
      beforeEach(() => {
        return createWrapper();
      });

      it('does not render the loading skeleton', () => {
        expect(findSkeletonLoaders(testId).length).toBe(0);
      });

      it('renders the metric values', () => {
        expect(findTableRow(testId).text()).toMatchSnapshot();
      });

      if (change === 0) {
        it('does not render the trend indicator', () => {
          expect(findTrendIndicator(testId).exists()).toBe(false);
        });
      } else {
        it('renders the trend indicator', () => {
          expect(findTrendIndicator(testId).props().change).toBe(change);
        });
      }

      it(`${hasValueTooltips ? 'adds' : 'does not add'} tooltip to value cells`, () => {
        const tooltip = findValueTableCells(testId).at(0).findComponent(GlTooltip);
        expect(tooltip.exists()).toBe(hasValueTooltips);
      });

      it(`${hasValueTooltips ? 'adds' : 'does not add'} hover classes to value cells`, () => {
        const hoverClasses = ['gl-cursor-pointer', 'hover:gl-underline'];
        const formattedMetric = findValueTableCells(testId)
          .at(0)
          .find('[data-testid="formatted-metric-value"]');

        expect(formattedMetric.classes().some((c) => hoverClasses.includes(c))).toBe(
          hasValueTooltips,
        );
      });
    });
  });

  describe('i18n', () => {
    describe.each`
      language   | formattedValue
      ${'en-US'} | ${'5,000'}
      ${'de-DE'} | ${'5.000'}
    `('When the language is $language', ({ formattedValue, language }) => {
      beforeEach(() => {
        setLanguage(language);
        return createLargeValuesWrapper();
      });

      it('formats numbers correctly', () => {
        expect(findTableRow('ai-impact-metric-vulnerability-critical').html()).toContain(
          formattedValue,
        );
      });
    });
  });
});
