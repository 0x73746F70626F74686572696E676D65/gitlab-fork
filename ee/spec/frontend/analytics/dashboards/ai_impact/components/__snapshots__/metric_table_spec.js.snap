// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metric table for the change_failure_rate table row when the data is loaded renders the metric values 1`] = `
"Change failure rate  
    Percentage of deployments that cause an incident in production.
    
      Go to docs
      
      10.0%
    
      20.0%
    
      40.0%
    
      10.0%
    
      20.0%
    
      40.0%
    
  100.0%"
`;

exports[`Metric table for the code_suggestions_usage_rate table row when the data is loaded renders the metric values 1`] = `
"Code Suggestions usage  
    Monthly user engagement with AI Code Suggestions. Percentage ratio calculated as monthly unique Code Suggestions users / total monthly unique code contributors.
    
      Go to docs
      
      5.0%
    
      10.0%
    
      25.0%
    
      5.0%
    
      10.0%
    
      25.0%
    
  100.0%"
`;

exports[`Metric table for the cycle_time table row when the data is loaded renders the metric values 1`] = `
"Cycle time  
    Median time from the earliest commit of a linked issue's merge request to when that issue is closed.
    
      Go to docs
      
      4.0 d
    
      2.0 d
    
      1.0 d
    
      4.0 d
    
      2.0 d
    
      1.0 d
    
  50.0%"
`;

exports[`Metric table for the deployment_frequency table row when the data is loaded renders the metric values 1`] = `
"Deployment frequency  
    Average number of deployments to production per day. This metric measures how often value is delivered to end users.
    
      Go to docs
      
      10.0/d
    
      20.0/d
    
      40.0/d
    
      10.0/d
    
      20.0/d
    
      40.0/d
    
  100.0%"
`;

exports[`Metric table for the lead_time table row when the data is loaded renders the metric values 1`] = `
"Lead time  
    Median time from issue created to issue closed.
    
      Go to docs
      
      -
    
      2.0 d
    
      4.0 d
    
      1.0 d
    
      2.0 d
    
      4.0 d"
`;

exports[`Metric table for the vulnerability_critical table row when the data is loaded renders the metric values 1`] = `
"Critical vulnerabilities over time  
    Critical vulnerabilities over time.
    
      Go to docs
      
      40
    
      20
    
      10
    
      40
    
      20
    
      10
    
  50.0%"
`;
